src/Perl6/Metamodel/Concretization.nqp

```diff
@@ -35,25 +37,31 @@ role Perl6::Metamodel::Concretization {
         @conc
     }
 
-    method !rebuild_table() {
-        my %new_conc_table := nqp::clone(%!conc_table);
-        for @!concretizations {
-            %new_conc_table{~nqp::objectid(nqp::decont($_[0]))} := nqp::decont($_[1]);
-        }
-        nqp::scwbdisable();
-        %!conc_table := %new_conc_table;
-        nqp::scwbenable();
+    method !maybe_rebuild_table() {
+        my %new_conc_table;
+        $lock.protect: {
+            %new_conc_table := nqp::clone(%!conc_table);
```

Comment:
We can do a micro-optimization here by doiing initial binding of `%new_conc_table` to `%!conc_table` and moving `clone` inside the following `if`. Since we never modify the content of the hash already bound to the attribute it is safe to return it without cloning if no changes are needed. And only clone when there are new entries.